Always fetch missing CellML files

**Description**
`FetchContent` currently skips downloading CellML files if configuration has already been done once. This fix was added in #26 and prevents `FetchContent` from throwing errors when it can't overwrite existing files, which can happen when `cmake` is called twice or `ccmake` is used. 

The current behaviour means that CellML files which have been accidentally removed after the inital configuration will not be replaced when `cmake` is called again.

**Steps To Reproduce**
```
cd /path/to/Chaste/build
cmake ..
rm /path/to/Chaste/projects/ApPredict/src/cellml
cmake ..
make ApPredict
```

**Error**
```
[ 82%] Building CXX object projects/ApPredict/CMakeFiles/chaste_project_ApPredict.dir/src/stats/BayesianInferer.cpp.o
/home/runner/_work/ApPredict/ApPredict/Chaste/projects/ApPredict/src/single_cell/CipaQNetCalculator.cpp:42:10: fatal error: ohara_rudy_cipa_v1_2017Cvode.hpp: No such file or directory
   42 | #include "ohara_rudy_cipa_v1_2017Cvode.hpp"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[3]: *** [projects/ApPredict/CMakeFiles/chaste_project_ApPredict.dir/build.make:258: projects/ApPredict/CMakeFiles/chaste_project_ApPredict.dir/src/single_cell/CipaQNetCalculator.cpp.o] Error 1
make[3]: *** Waiting for unfinished jobs....
[ 82%] Building CXX object projects/ApPredict/CMakeFiles/chaste_project_ApPredict.dir/src/stats/LinearDiscriminantAnalysis.cpp.o
make[2]: *** [CMakeFiles/Makefile2:22324: projects/ApPredict/CMakeFiles/chaste_project_ApPredict.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:2327: CMakeFiles/project_ApPredict.dir/rule] Error 2
make: *** [Makefile:314: project_ApPredict] Error 2
```


**To Do**
* Try to force overwrites (delete and re-fetch), or check for missing files and fetch only those.

--- a/.github/workflows/ubuntu-2004.yml
+++ b/.github/workflows/ubuntu-2004.yml
@@ -56,6 +56,14 @@ jobs:
     - name: cmake configure
       run: nice -n 10 cmake -DCMAKE_BUILD_TYPE=Release ..
       working-directory: Chaste/build
+      
+    - name: remove cellml
+      run: rm -rf ApPredict/src/cellml
+      working-directory: Chaste/projects
+      
+    - name: cmake configure again
+      run: nice -n 10 cmake -DCMAKE_BUILD_TYPE=Release ..
+      working-directory: Chaste/build
 
     - name: build core libraries
       run: nice -n 10 cmake --build . --parallel ${NPROC} --target chaste_core

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -43,7 +43,7 @@
 # Note that the order in which components are specified does not matter.
 
 set(ApPredict_CELLML_TAG e394aba8a7ca50f72ee212e21431045eae8b5ca4)
-set(ApPredict_CELLML_URL "https://github.com/Chaste/cellml/blob/${ApPredict_CELLML_TAG}/cellml")
+set(ApPredict_CELLML_URL "https://raw.githubusercontent.com/Chaste/cellml/${ApPredict_CELLML_TAG}/cellml")
 set(ApPredict_CELLML_SRC "${CMAKE_CURRENT_SOURCE_DIR}/src/cellml")
 set(ApPredict_CELLML_FILES
     davies_isap_2012.cellml
@@ -65,11 +65,17 @@ message (STATUS "Downloading ApPredict cellml files to ${ApPredict_CELLML_SRC}")
 foreach(cellml_file ${ApPredict_CELLML_FILES})
     if (NOT EXISTS "${ApPredict_CELLML_SRC}/${cellml_file}")
         file(DOWNLOAD
-            "${ApPredict_CELLML_URL}/cellml/${cellml_file}"
+            "${ApPredict_CELLML_URL}/${cellml_file}"
             "${ApPredict_CELLML_SRC}/${cellml_file}"
-            TIMEOUT 5
+            TIMEOUT 10
             TLS_VERIFY ON
+            STATUS dl_status
         )
+        list (GET dl_status 0 dl_status_code)
+        if (NOT dl_status_code EQUAL 0)
+            file(REMOVE "${ApPredict_CELLML_SRC}/${cellml_file}") # remove stub
+            message (WARNING "Could not download ${cellml_file}")
+        endif ()
     endif ()
 endforeach()
 

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -42,36 +42,35 @@
 #
 # Note that the order in which components are specified does not matter.
 
-include(FetchContent)
-message (STATUS "Downloading ApPredict cellml files to ${CMAKE_CURRENT_SOURCE_DIR}/src/cellml")
-Set(FETCHCONTENT_UPDATES_DISCONNECTED ON)
-FetchContent_Declare(CellML_repo_appredict
-                     GIT_REPOSITORY https://github.com/Chaste/cellml.git
-                     GIT_TAG e394aba8a7ca50f72ee212e21431045eae8b5ca4
-                     SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/src/cellml)
-FetchContent_MakeAvailable(CellML_repo_appredict)
+set(ApPredict_CELLML_TAG e394aba8a7ca50f72ee212e21431045eae8b5ca4)
+set(ApPredict_CELLML_URL "https://github.com/Chaste/cellml/blob/${ApPredict_CELLML_TAG}/cellml")
+set(ApPredict_CELLML_SRC "${CMAKE_CURRENT_SOURCE_DIR}/src/cellml")
+set(ApPredict_CELLML_FILES
+    davies_isap_2012.cellml
+    decker_2009.cellml
+    faber_rudy_2000.cellml
+    grandi_pasqualini_bers_2010_ss.cellml
+    hund_rudy_2004.cellml
+    livshitz_rudy_2007.cellml
+    mahajan_shiferaw_2008.cellml
+    ohara_rudy_2011_endo.cellml
+    ohara_rudy_cipa_v1_2017.cellml
+    paci_hyttinen_aaltosetala_severi_ventricularVersion.cellml
+    priebe_beuckelmann_1998.cellml
+    shannon_wang_puglisi_weber_bers_2004.cellml
+    ten_tusscher_model_2006_epi.cellml
+)
 
-# remove .git folder to prevent accidentally checking in deletion of cellml files
-file(REMOVE_RECURSE ${CMAKE_CURRENT_SOURCE_DIR}/src/cellml/.git)
-
-# remove cellml files that are not required
-file(GLOB cellml_files ${CMAKE_CURRENT_SOURCE_DIR}/src/cellml/cellml/*.cellml)
-foreach(cellml_file ${cellml_files})
-    if (NOT ${cellml_file} MATCHES davies_isap_2012.cellml
-        AND NOT ${cellml_file} MATCHES decker_2009.cellml
-        AND NOT ${cellml_file} MATCHES faber_rudy_2000.cellml
-        AND NOT ${cellml_file} MATCHES grandi_pasqualini_bers_2010_ss.cellml
-        AND NOT ${cellml_file} MATCHES hund_rudy_2004.cellml
-        AND NOT ${cellml_file} MATCHES livshitz_rudy_2007.cellml
-        AND NOT ${cellml_file} MATCHES mahajan_shiferaw_2008.cellml
-        AND NOT ${cellml_file} MATCHES ohara_rudy_2011_endo.cellml
-        AND NOT ${cellml_file} MATCHES ohara_rudy_cipa_v1_2017.cellml
-        AND NOT ${cellml_file} MATCHES paci_hyttinen_aaltosetala_severi_ventricularVersion.cellml
-        AND NOT ${cellml_file} MATCHES priebe_beuckelmann_1998.cellml
-        AND NOT ${cellml_file} MATCHES shannon_wang_puglisi_weber_bers_2004.cellml
-        AND NOT ${cellml_file} MATCHES ten_tusscher_model_2006_epi.cellml)
-        file(REMOVE ${cellml_file})
-    endif()
+message (STATUS "Downloading ApPredict cellml files to ${ApPredict_CELLML_SRC}")
+foreach(cellml_file ${ApPredict_CELLML_FILES})
+    if (NOT EXISTS "${ApPredict_CELLML_SRC}/${cellml_file}")
+        file(DOWNLOAD
+            "${ApPredict_CELLML_URL}/cellml/${cellml_file}"
+            "${ApPredict_CELLML_SRC}/${cellml_file}"
+            TIMEOUT 5
+            TLS_VERIFY ON
+        )
+    endif ()
 endforeach()
 
 # Here we just depend on core components (nothing application-specific).

--- a/README.md
+++ b/README.md
@@ -22,6 +22,6 @@ so that all the file paths can be picked up correctly (replacing ```<chaste sour
 
 This ApPredict project should be used with the current `develop` branch of [Chaste](https://github.com/Chaste/Chaste). If instead you want a version that works with a released version of Chaste, then please select the relevant Tag of this github repository.
 
-(In the past we had a git submodule with CellML files from www.github.com/Chaste/CellML, but this is now pulled in by a cmake [fetch](https://cmake.org/cmake/help/latest/module/FetchContent.html))
+(In the past we had a git submodule with CellML files from www.github.com/Chaste/CellML, but this is now pulled in by a cmake [file download](https://cmake.org/cmake/help/latest/command/file.html#download))
 
 [Getting Started]: <https://chaste.cs.ox.ac.uk/trac/wiki/GettingStarted>

--- a/src/extra_models/README
+++ b/src/extra_models/README
@@ -3,7 +3,7 @@ Users can put additional cellml models in this directory.
 Once the project is built these will be available as pre-compiled models 
 and can be used with --model <name of cellml file (without.cellml)>.
 
-Please note: The additional cellml file cannot have the same name as one of the pre-included cellml files in src/cellml/cellml
+Please note: The additional cellml file cannot have the same name as one of the pre-included cellml files in src/cellml
 as these would generate model code with conflicting names. 
 In case you receive a compile error with regards to duplicate definitions, 
 we suggest renaming the additional cellml files to avoid conflicts.

--- a/test/TestApPredict.hpp
+++ b/test/TestApPredict.hpp
@@ -112,7 +112,7 @@ public:
                                   "No model matches this index: 99999");
         }
         {
-            CommandLineArgumentsMocker wrapper("--cellml projects/ApPredict/src/cellml/cellml/ten_tusscher_model_2006_epi.cellml --plasma-concs 1 10 --pic50-herg 4.5 --plasma-conc-logscale false --output-dir ApPredict_output_long");
+            CommandLineArgumentsMocker wrapper("--cellml projects/ApPredict/src/cellml/ten_tusscher_model_2006_epi.cellml --plasma-concs 1 10 --pic50-herg 4.5 --plasma-conc-logscale false --output-dir ApPredict_output_long");
 
             ApPredictMethods methods;
             TS_ASSERT_EQUALS(Warnings::Instance()->GetNextWarningMessage(), "Argument --cellml <file> is deprecated: use --model <file> instead.");

--- a/test/TestApPredictLong.hpp
+++ b/test/TestApPredictLong.hpp
@@ -64,12 +64,12 @@ public:
         // loop over hardcoded and dynamically loaded.
         // try both --model --cellml and --cellml and try both relative and absolute paths
 
-        FileFinder cellm_file("projects/ApPredict/src/cellml/cellml/ten_tusscher_model_2006_epi.cellml", RelativeTo::CWD);
+        FileFinder cellm_file("projects/ApPredict/src/cellml/ten_tusscher_model_2006_epi.cellml", RelativeTo::CWD);
         std::vector<std::string> model_args = { "--model 2",
                                                 "--model ten_tusscher_model_2006_epi",
-                                                "--model projects/ApPredict/src/cellml/cellml/ten_tusscher_model_2006_epi.cellml",
+                                                "--model projects/ApPredict/src/cellml/ten_tusscher_model_2006_epi.cellml",
                                                 "--model " + cellm_file.GetAbsolutePath(),
-                                                "--cellml projects/ApPredict/src/cellml/cellml/ten_tusscher_model_2006_epi.cellml"};
+                                                "--cellml projects/ApPredict/src/cellml/ten_tusscher_model_2006_epi.cellml"};
         for (unsigned i = 0; i < model_args.size(); i++)
         {
             CommandLineArgumentsMocker wrapper(model_args[i] + " --plasma-concs 1 10 --pic50-herg 4.5 --plasma-conc-logscale false --output-dir ApPredict_output_long");

